File: Test 2 - Master mind/2.py
import random

def generate_pattern(length):
    """Gera um padrão aleatório de números únicos."""
    return random.sample("123456", length)

def check_guess(pattern, guess):
    """
    Compara o palpite com o padrão.
    Retorna uma tupla: (numeros_corretos, posicoes_corretas)
    """
    correct_position = 0
    for i in range(len(pattern)):
        if guess[i] == pattern[i]:
            correct_position += 1
    
    # Como o padrão não tem repetições, usamos conjuntos para achar números corretos
    correct_numbers = len(set(guess).intersection(set(pattern)))
    
    return correct_numbers, correct_position

def main():
    # 1. Configuração do Jogo: Tamanho do Padrão
    while True:
        try:
            length_input = input("Enter pattern length (1-6): ")
            pattern_length = int(length_input)
            if 1 <= pattern_length <= 6:
                break
            print("Invalid length. Please enter a number between 1 and 6.")
        except ValueError:
            print("Invalid length. Please enter a number between 1 and 6.")

    # 2. Configuração do Jogo: Número de Tentativas
    while True:
        try:
            guesses_input = input("Enter number of guesses: ")
            max_guesses = int(guesses_input)
            if max_guesses > 0:
                break
            print("Invalid number. Please enter a number greater than 0.")
        except ValueError:
            print("Invalid number. Please enter a number greater than 0.")

    # Gera o padrão (números únicos de 1 a 6)
    pattern = generate_pattern(pattern_length)
    
    current_attempt = 0

    # Loop Principal do Jogo
    while current_attempt < max_guesses:
        guess_str = input("Type your guess: \n")

        # Validações de entrada
        if len(guess_str) != pattern_length:
            print(f"Input needs to be {pattern_length} of size.")
            continue
        
        if not all(c in "123456" for c in guess_str):
            print("Try something like a number between 1 and 6.")
            continue

        current_attempt += 1
        guess = list(guess_str)

        # Chama a função pura para processar a lógica
        correct_numbers, correct_position = check_guess(pattern, guess)

        # Formatação da saída (pluralização correta)
        msg_num = "number" if correct_numbers == 1 else "numbers"
        msg_pos = "number" if correct_position == 1 else "numbers"
        
        print(f"{correct_numbers} {msg_num} correct, {correct_position} {msg_pos} in correct position")

        # Verifica vitória
        if correct_position == pattern_length:
            print(f"You broke the code in {current_attempt} guesses!")
            return

    # Fim de jogo (Derrota)
    print(f"You were unable to break the code in {max_guesses} guesses. Code pattern is: {''.join(pattern)}.")

if __name__ == "__main__":
    main()

File: Test 2 - Master mind/2.js
var readline = require('readline');
var maxAttempt = null;
var currentAttempt = 0;
var generatedNumber = [];
var patternLength = 0;

var promptRead = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

startGame();

function startGame() {
    // 1. Pergunta o tamanho do padrão (Pattern Length)
    promptRead.question('Enter pattern length (1-6): ', function (input) {
        let length = parseInt(input);
        if (isNaN(length) || length < 1 || length > 6) {
            console.log("Invalid length. Please enter a number between 1 and 6.");
            startGame();
            return;
        }
        patternLength = length;
        askGuesses();
    });
}

function askGuesses() {
    // 2. Pergunta o número de tentativas (Number of Guesses)
    promptRead.question('Enter number of guesses: ', function (input) {
        let guesses = parseInt(input);
        if (isNaN(guesses) || guesses < 1) {
            console.log("Invalid number. Please enter a number greater than 0.");
            askGuesses();
            return;
        }
        maxAttempt = guesses;
        callGuessPrompt();
    });
}

function callGuessPrompt() {
    promptRead.setPrompt('Type your guess: \n');
    promptRead.prompt();

    //Keep reading the input until something break stops the prompt.
    promptRead.on('line', function (input) {
        handleGuessPrompt(input);
        promptRead.prompt();
    });
}

function handleGuessPrompt(userInput) {
    if (!generatedNumber.length) {
        generatedNumber = generatesNumber();
    }
    handleInput(userInput);
}

function handleInput(userInput) {
    let userGuess = userInput.split('');
    let correctNumber = 0;
    let correctPosition = 0;

    //If the user sets the wrong amount of numbers, return.
    if (userGuess.length != patternLength) {
        console.log(`Input needs to be ${patternLength} of size.`);
        return;
    }

    //If any value outside 1...6
    if (userGuess.some(typedValue => typedValue > 6 || typedValue < 1 || isNaN(typedValue))) {
        console.log(`Try something like a number between 1 and 6.`);
        return;
    }
    //It has to be here. We just want to count the attempt if it's valid.
    currentAttempt++;

    // Calculate correctPosition
    userGuess.forEach((item, index) => {
        if (item === generatedNumber[index]) {
            correctPosition++;
        }
    });

    // Calculate correctNumber (Total unique numbers from guess found in pattern)
    // Using Set to avoid overcounting duplicates in guess (e.g. guess 1123 vs pattern 1234)
    const uniqueGuess = [...new Set(userGuess)];
    uniqueGuess.forEach((item) => {
        if (generatedNumber.includes(item)) {
            correctNumber++;
        }
    });

    console.log(
        `${correctNumber} number${correctNumber > 1 ? 's' : ''} correct, ${correctPosition} number${correctPosition > 1 ? 's' : ''} in correct position`
    );



    //If you want to see how it's working, you could uncomment these lines.
    /* 
    console.log("userGuess")
    console.log(userGuess)
    console.log("=======================")
    console.log("generatedNumber")
    console.log(generatedNumber)
    console.log("=======================")
    console.log(correctNumber)
    console.log(correctPosition)
 */
    if (
        correctNumber == patternLength &&
        correctPosition == patternLength
    ) {
        console.log(`You broke the code in ${currentAttempt} guesses!`);
        promptRead.close();
        process.exit(0);
    } else if (currentAttempt == parseInt(maxAttempt)) {
        console.log(`You were unable to break the code in ${currentAttempt} guesses. Code pattern is: ${generatedNumber.join('')}.`);
        promptRead.close();
        process.exit(0);
    }
}

//Generates game array numbers
function generatesNumber() {
    let number = [];
    for (let index = 0; index < patternLength;) {
        let numberGenerated = parseInt(Math.random() * 6) + 1;
        if (!number.includes(numberGenerated.toString())) {
            number.push(numberGenerated.toString());
            index++;
        }
    }
    return number;
}